### PR TITLE
temp: Automatic Ticket Revocations Proposal.

### PR DESCRIPTION
**Pull request by rstaudt2**
Draft proposal for Automatic Ticket Revocations Consensus Change.

### CODE REVIEW

File: temp/automatic_ticket_revocations_proposal.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+This proposal intends to significantly improve the stakeholder user experience around ticket revocations by automatically revoking missed or expired tickets.
+
+Further, it will enable the recovery of funds for users who lost their redeem script for the legacy VSP system (before the release of [vspd](https://blog.decred.org/2020/06/02/A-More-Private-Way-to-Stake/), which removed the need for the redeem script).
+
+## Current State of Ticket Revocations
+
+Missed and expired tickets must be revoked to return the funds used to purchase the ticket to the user.  Missed and expired tickets are defined as follows:
+
+- **Missed tickets** are tickets that fail to cast a vote when selected.  Tickets are pseudorandomly selected to vote for each block.  If the wallet with voting rights for that ticket is not online to cast its vote when a ticket is selected to vote, the ticket is marked as missed.
```
Comment (davecgh) [inline temp/automatic_ticket_revocations_proposal.md]: I would probably reword this a bit to account for the possibility that votes can be cast and still not included under some circumstances, such as taking too long to vote.  For example, a PoW miner might already be working on a solution to a block when the vote is cast and the miner finds the solution before generating a new template.  The details here can get pretty technical, and I don't think it makes sense to try and characterize them all here as it's mostly tangential to the DCP, but I do think it should be accurate, and the current text is only really partially correct.  Perhaps a tweak such as:

```suggestion
- **Missed tickets** are tickets that do not have a corresponding vote included in the block they are selected.  Tickets are pseudorandomly selected to vote for each block.  The most common cause of missed tickets occurs when a wallet with voting rights for that ticket is not online to cast its vote when a ticket is selected to vote.
```

File: temp/automatic_ticket_revocations_proposal.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+- **Expired tickets** are tickets that are not selected to vote before the ticket expiration duration (40,960 blocks, or approximately 142 days, on mainnet).
+
+Revoking a missed or expired ticket requires broadcasting a **ticket revocation transaction**.  In order to manually revoke a ticket today, a user needs to:
+- Open the wallet that has voting rights for the ticket so that the ticket revocation transaction can be signed
+- If the wallet is running in SPV mode, the user will need to disable SPV, sync the blockchain, and rescan
```
Comment (davecgh) [inline temp/automatic_ticket_revocations_proposal.md]: This isn't strictly true because it's not really required.  It is correct that the current release version of Decrediton requires this, but that is because there isn't a good way for it to know when a ticket has been missed in order to only conditionally show the revoke button when one or more tickets are missed.  However, there is no reason it has to work that way, and, I'm actually fairly certain next Decrediton release is changing this by adding a button which will just have no effect when a ticket is not eligible to be revoked.

In other words, I would probably remove this as it isn't actually a requirement imposed by the existing consensus rules.  Alternatively, you could include something similar, but if you decide to, I think it should call out the actual root cause stemming from the consensus rules and the resulting poor user experience that occurs as a result.

File: temp/automatic_ticket_revocations_proposal.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+This proposal proposes the following changes to enable automatic ticket revocations:
+- Allow tickets to be revoked by anyone instead of requiring a valid signature from the voting address
+	- Ticket revocations will still have to go to the original commitment address in the ticket, as they do today, so this does not allow theft of funds
+	- This would not apply to votes, and therefore voting rights for tickets would remain as they are today
+- Require blocks to contain ticket revocation transactions for all tickets that are missed or expired as of that block
+
+## Consensus Rule Change Cost-Benefit Analysis
+
+When considering changes to the consensus rules, it is paramount to examine both the positive and negative consequences of the proposed changes.  There are nearly always tradeoffs involved in one form or another.  Since there are no known agreed-upon cons to these changes, the following analysis contains one section for pros and another for an expected potential objection.
```
Comment (davecgh) [inline temp/automatic_ticket_revocations_proposal.md]: There are some additional potential cons I can think of, depending on whose point of view you're considering it from:

- It removes fees from revocations which slightly reduces potential income to PoW miners
- All wallet implementation will need to be updated to accommodate the proposed changes which involves both back-end modifications as well as front-end design work
- All custom mining template generation outside of dcrd will also have to be updated to properly include the required revocations which might be a non-trivial amount of work

I would also suggest a bit of a recap after with a brief discussion addressing those cons, because, a couple of them are a matter of perspective.  e.g. stake holders not having to pay another transaction fee after already having their ticket expire would likely be welcomed, while PoW miners would prefer the additional fees.

File: temp/automatic_ticket_revocations_proposal.md (hunk 1 above)
Comment (rstaudt2) [reply temp/automatic_ticket_revocations_proposal.md]: Good point, updated with your suggestion to account for that.

File: temp/automatic_ticket_revocations_proposal.md (hunk 2 above)
Comment (rstaudt2) [reply temp/automatic_ticket_revocations_proposal.md]: Ya, good point that it only applies to the current version of Decredition.  I just went ahead and removed it as suggested.

File: temp/automatic_ticket_revocations_proposal.md (hunk 3 above)
Comment (rstaudt2) [reply temp/automatic_ticket_revocations_proposal.md]: Thanks for outlining these!  I've updated this section to describe the various cons as you recommended.  I think I was originally not considering the work outside of `dcrd` a con and more of just required work that comes with accommodating any consensus change, but it makes sense and I agree that any additional required work should always be explicitly listed as a tradeoff to be considered.

File: temp/automatic_ticket_revocations_proposal.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+The general audience section provides a high-level overview of the proposal.  It includes details suitable for all stakeholders, including those unfamiliar with the inner workings of consensus.
+
+The technical details section contains a more in-depth discussion of the relevant technical aspects, implementation specifics, and rationale for the chosen semantics of the proposal.
+
+# General Audience
+
+## High-Level Overview
+
+This proposal intends to significantly improve the stakeholder user experience around ticket revocations by automatically revoking missed or expired tickets.
```
Comment (davecgh) [inline temp/automatic_ticket_revocations_proposal.md]: I might suggest the following slight tweak to the verbiage here to make it less ambiguous:

```suggestion
This proposal intends to significantly improve the stakeholder user experience around ticket revocations by removing the requirement for stakeholders to manually revoke missed and expired tickets.
```

I say that because, after thinking about it from the perspective of someone who is not already familiar with everything, which is the intent of this section, merely saying "by automatically revoking" seems a bit ambiguous to me. My first thoughts would be "Who or what is doing that automatic revocation?"  Finally, I would suggest ` and` over `or` since it's inclusive of both.

It's probably already reasonable enough with any changes, so it's obviously up to you if you think it's worth tweaking or not.

File: temp/automatic_ticket_revocations_proposal.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+Further, it will enable the recovery of funds for users who lost their redeem script for the legacy VSP system (before the release of [vspd](https://blog.decred.org/2020/06/02/A-More-Private-Way-to-Stake/), which removed the need for the redeem script).
+
+## Current State of Ticket Revocations
+
+Missed and expired tickets must be revoked to return the funds used to purchase the ticket to the user.  Missed and expired tickets are defined as follows:
```
Comment (davecgh) [inline temp/automatic_ticket_revocations_proposal.md]:
```suggestion
Missed and expired tickets must be revoked to return the funds used to purchase the ticket to the stakeholder.  Missed and expired tickets are defined as follows:
```

File: temp/automatic_ticket_revocations_proposal.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+- Allow tickets to be revoked by anyone instead of requiring a valid signature from the voting address
+	- Ticket revocations will still have to go to the original commitment address in the ticket, as they do today, so this does not allow theft of funds
+	- This would not apply to votes, and therefore voting rights for tickets would remain as they are today
+- Require blocks to contain ticket revocation transactions for all tickets that are missed or expired as of that block
+
+## Consensus Rule Change Cost-Benefit Analysis
+
+When considering changes to the consensus rules, it is paramount to examine both the positive and negative consequences of the proposed changes.  There are nearly always tradeoffs involved in one form or another.  The following two sections provide an analysis in the form of pros and cons accordingly.
+
+### Pros
```
Comment (davecgh) [inline temp/automatic_ticket_revocations_proposal.md:44]: I would suggest an addition pro here:

- It simplifies wallet implementations since they will no longer have to include logic to detect the revocation conditions nor create and broadcast revocation transactions in response to them

**Review by davecgh — COMMENTED** [review]
The updates look good.  I have a few more minor suggestions from another pass.

File: temp/automatic_ticket_revocations_proposal.md (hunk 4 above)
Comment (rstaudt2) [reply temp/automatic_ticket_revocations_proposal.md]: I think that is a good suggestion, updated.

File: temp/automatic_ticket_revocations_proposal.md (hunk 6 above)
Comment (rstaudt2) [reply temp/automatic_ticket_revocations_proposal.md:44]: That is a good pro to call out.  Thanks for the suggestion!

**Review by davecgh — COMMENTED** [review]
This is good to go from my perspective.  I don't see anything else.

**Comment by rstaudt2** [comment]
> This is good to go from my perspective. I don't see anything else.

OK great.  Thanks again for reviewing, I appreciate the feedback prior to submitting.
